doc/man3/SSL_set_async_callback.pod

```diff
@@ -0,0 +1,82 @@
+=pod
+
+=head1 NAME
+
+SSL_set_async_callback,
+SSL_get_async_status,
+SSL_clear_async_status
+- manage asynchronous operations
+
+=head1 SYNOPSIS
+
+=for comment multiple includes
+
+ #include <openssl/ssl.h>
+
+ int SSL_set_async_callback(SSL *s, ASYNC_callback_fn callback,
```

Comment:
Should there be an SSL_CTX version of this?

Reply:
the callback can be used in async only, and if SSL_CTX has "ASYNC_WAIT_CTX" field, then we need to add it. Currently it seems no.

Reply:
SSL_CTX is a factory for creating SSL objects. So if you want to be able to set it on *all* the SSL objects created from a single SSL_CTX you should be able to do so, rather than having to do it on each individual one.

Reply:
And inherit the callback function pointer to the SSL objects that are created from the SSL_CTX?

Reply:
If so, the callback name in the SSL_CTX structure could be prefixed with a 'default_' wording and it's only used to initialize the child SSL objects spawned from the SSL_CTX. And the user can still choose to override the default callback in the SSL object obtained from birth (Just like the certificate verification callback)

Reply:
This is a common pattern in SSL_CTX (i.e. copying values from SSL_CTX to SSL). Normally the name of the variable is the same in the SSL_CTX as it is in SSL

Reply:
Hmmm, I found this a little complicated. Say, if someone assign a callback to an SSL_CTX and then uses the SSL_CTX to create a bundle of SSL objects. When the SSL objects are being created, since the async callback is not NULL in the parent SSL_CTX object, so every SSL object should be initialized by calling `SSL_set_async_callback` during this creation (for instance in the `SSL_new()` API). Makes sense?

Reply:
It's not expected to alloc memory when setting the callback to SSL_CTX, but do that allocation when SSL is created. Sounds like this is the proper way of implementing this

Reply:
Right - that makes sense to me.

Reply:
Hmmm, if the callback argument is SSL specified, I think setting the callback function pointer to the parent level SSL_CTX is meaningless...

Reply:
yes, agree

Reply:
I'm still not comfortable with this answer. It also points out to me something else which is strange about this design: The callback that we specify on the SSL object does not get passed back the SSL object itself, i.e. I might expect to see an SSL specific callback that looks something like:

````
typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
````
This would be more consistent with other SSL callbacks, and probably removes the need for an SSL specific arg (meaning we could just set the callback at the `SSL_CTX` level).

Reply:
Another choice could be separating the `ASYNC allocating/setting` logic from `SSL_set_async_callback` by introducing a new API like `SSL_async_init`, in which the `ASYNC allocating/setting` works are done. Then the calling sequence could be:

1. SSL_set_async_callback or SSL_CTX_set_async_callback
2. SSL_async_init (using callback and args set in step 1)

Reply:
just come back from travel, and will update this PR in this week to address your comments

Reply:
Thanks man and I guess @mattcaswell would consider this is more proper design ;-)

Reply:
Hi, Matt,
I understand your points to make  _SSL_set_async_callback_ to be similar implementation like other SSL callback, but actually this callback is not in SSL level, but crypto level. See detailed info below:

1) Current asynchronous code is **not** in SSL layer, but in crypto layer, and the design objective is to make an asynchronous API call for HW . Asynchronous API requires a communication mechanism for application code to sync up with HW, thus a communication mechanism is provided.  Like current "fd" based mechanism, the async callback is to achieve SSL independence. Thus we design the callback below.

typedef int (*ASYNC_callback_fn)(void *arg);

Above callback definition is not in ssl.h, but in async.h, and it can work with libcrypto only, and independent of SSL. 

2) When handling SSL, the ASYNC_WAIT_CTX data structure is hidden in public SSL data structure. We could not access the data structure directly in SSL. To resolve this, the existing fd based mechanism provides some API to allow user to access this field in SSL context. 

___owur int SSL_get_all_async_fds(SSL *s, OSSL_ASYNC_FD *fds, size_t *numfds);_

3) The current design of SSL_set_async_callback works the similar way like above fd based mechanism.

4) Maybe above _SSL_set_async_callback_ is bit confusing, and this API is actually to set crypto related callback by updating aASYNC_WAIT_CTX data structure. It is actually SSL_set_**crypto**_async_callback. :)

5) I am also trying to add new API in SSL layer to address your question, but seemly we don't have SSL layer async callback, and just "**crypto**" async implemented, and so there is no meaning to add "SSL" into the callback. My understanding of the callbacks under SSL_CTX is those callbacks which will be triggered in some SSL processing context. For example,

     /*-
         * For a server, this contains a callback function that allows the
         * server to select the protocol for the connection.
         *   out: on successful return, this must point to the raw protocol
         *        name (without the length prefix).
         *   outlen: on successful return, this contains the length of |*out|.
         *   in: points to the client's list of supported protocols in
         *       wire-format.
         *   inlen: the length of |in|.
         */
    int (*alpn_select_cb) (SSL *s,
                           const unsigned char **out,
                           unsigned char *outlen,
                           const unsigned char *in,
                           unsigned int inlen, void *arg);

In this case,  I fully understand the "SSL" object should be added. 
But in our case, the callback is not in SSL context, and can be triggered when **crypto** engine completed its job. 

6) In a summary, current "async" is crypto level, and all implementation is actually in async.h, and SSL level API is just allow SSL application to access the hidden data structure. I can rename the SSL API from  _SSL_set_async_callback()_ to _SSL_set_crypto_async_callback()_, so we reserve it for real SSL level async in the future. 

Does it make sense?

Reply:
I'm still not satisfied with this answer. It is true that that the ASYNC api is implemented at the libcrypto level and therefore the callback at this level has no knowledge of SSL. My suggestion was that we introduce an SSL specific callback in libssl.

So, an application would call `SSL_set_async_callback()` or `SSL_CTX_set_async_callback()` with a callback that looks like this:

````
typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
````

Internally libssl would call `ASYNC_WAIT_CTX_set_callback()` with its own internal callback implementation. Something like:

````
int SSL_set_async_callback_arg(SSL *s, void *arg)
{
    s->async_cb_arg = arg;
    return 1;
}

int SSL_set_async_callback(SSL *s, SSL_async_callback_fn callback)
{
    s->async_cb = callback;
    return 1;
}

...

static int ssl_async_wait_ctx_cb(void *arg)
{
    SSL *s = (SSL *)arg;

    return s->async_cb(s, s->async_cb_arg);
}

static int ssl_start_async_job(SSL *s, struct ssl_async_args *args,
                               int (*func) (void *))
{
    int ret;
    if (s->waitctx == NULL) {
        s->waitctx = ASYNC_WAIT_CTX_new();
        if (s->waitctx == NULL)
            return -1;
        if (!ASYNC_WAIT_CTX_set_callback(s->waitctx, ssl_async_wait_ctx_cb, s))
            return -1;
    }
    ...
}
````

Reply:
very great comment. Will update PR according to your idea.

Reply:
I prefer having two separate calls at both SSL_CTX and SSL layers because it provides more flexibility. Actually it might not even be necessary to set an arg at all for some applications (perhaps the SSL object is enough)

Reply:
Why doe we need `SSL_CALLBACK_MODE` here? Isn't it sufficient to test whether `s->async_cb` is NULL or not?

Reply:
yeah, you are right. Will update it.